src/main/kotlin/com/criptext/mail/services/DecryptionService.kt

```diff
@@ -0,0 +1,187 @@
+                                filesDir = applicationContext.filesDir,
+                                cacheDir = applicationContext.cacheDir),
+                        host = this,
+                        isPostNougat = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N,
+                        activeAccount = account,
+                        storage = storage)
+            }
+        }
+        val pushData = data as Map<String, String>
+        if(pushData.isNotEmpty()) {
+            val shouldPostNotification = !MessagingService.isAppOnForeground(applicationContext, applicationContext.packageName)
+            pushController?.parsePushPayload(pushData, shouldPostNotification)
+        } else {
+            pushController = null
+        }
+    }
+
+    private fun stopService() {
+        // Stop foreground service and remove the notification.
+        stopForeground(true)
+        // Stop the foreground service.
+        stopSelf()
+
+        IS_RUNNING = false
+    }
+
+
+    override fun onDestroy() {
+        IS_RUNNING = false
+    }
+
+    private fun prepareNotification(){
+        val notification = NotificationCompat.Builder(applicationContext, channelId)
+                .setSmallIcon(R.mipmap.ic_launcher)
+                .setContentTitle("Decrypting Emails")
+                .build()
+        startForeground(NOTIFICATION_ID, notification)
+        IS_RUNNING = true
+    }
+
+    private fun createNotificationChannel() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            var notificationChannel =
+                    notificationManager?.getNotificationChannel(channelId)
+            if (notificationChannel == null) {
+                notificationChannel = NotificationChannel(
+                        channelId, TAG, NotificationManager.IMPORTANCE_LOW
+                )
+                notificationManager?.createNotificationChannel(notificationChannel)
+            }
+        }
+    }
+
+    fun pushProcessed(){
+        if(queue.isEmpty()) {
+            pushController?.doGetEvents()
+        } else {
+            data = queue.first()
+            queue.removeAt(0)
+            getPushNotification()
+        }
+    }
+
+    fun endService(){
+        pushController = null
+        stopService()
+    }
+
+    fun notifyPushEvent(notifier: Notifier?){
+        notifier?.notifyPushEvent(applicationContext)
+    }
+
+    fun cancelPush(notificationId: Int, storage: KeyValueStorage, key: KeyValueStorage.StringKey, headerId: Int){
+        val manager = applicationContext
+                .getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+        val notCount = storage.getInt(key, 0)
+        manager.cancel(notificationId)
+        if((notCount - 1) <= 0) {
+            manager.cancel(headerId)
+        }
+        storage.putInt(key, if(notCount <= 0) 0 else notCount - 1)
+    }
+
+    companion object {
+        const val NOTIFICATION_ID = 42
+        const val TAG = "ForegroundWorker"
```

Comment:
Change this to something more recognizable. This is visible in the notification settings of the app.